File: modules/agent/src/services/task_service.py
"""Service layer for task business logic."""

import logging
from typing import Dict, List, Optional, Any, Tuple
from fastapi import HTTPException
from repositories.task_repository import (
    find_all_tasks,
    find_task_by_id,
    find_tasks_by_entity,
    create_task as create_task_repo,
    update_task as update_task_repo,
    delete_task as delete_task_repo,
    get_entity_display_name,
    find_task_statuses,
    find_task_priorities,
    get_lead_type,
    get_account_entity,
    batch_get_entity_display_names,
    batch_get_lead_types,
    batch_get_account_entities,
    TaskCreate,
    TaskUpdate,
)

# Re-export Pydantic models for controllers
__all__ = ["TaskCreate", "TaskUpdate"]

logger = logging.getLogger(__name__)


async def get_tasks_paginated(
    tenant_id: int,
    project_id: Optional[int] = None,
    page: int = 1,
    page_size: int = 20,
    order_by: str = "created_at",
    order: str = "DESC",
    search: Optional[str] = None,
) -> Tuple[List[Any], int]:
    """Get tasks with pagination and optional project scope."""
    return await find_all_tasks(
        tenant_id=tenant_id,
        project_id=project_id,
        page=page,
        page_size=page_size,
        order_by=order_by,
        order=order,
        search=search,
    )


async def get_task(task_id: str) -> Any:
    """Get a task by ID."""
    try:
        task_id_int = int(task_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid task ID format")

    task = await find_task_by_id(task_id_int)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    return task


async def create_task(task_data: Dict[str, Any]) -> Any:
    """Create a new task."""
    return await create_task_repo(task_data)


async def update_task(task_id: str, task_data: Dict[str, Any]) -> Any:
    """Update a task."""
    try:
        task_id_int = int(task_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid task ID format")

    task = await update_task_repo(task_id_int, task_data)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    return task


async def delete_task(task_id: str) -> bool:
    """Delete a task."""
    try:
        task_id_int = int(task_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid task ID format")

    deleted = await delete_task_repo(task_id_int)
    if not deleted:
        raise HTTPException(status_code=404, detail="Task not found")

    return deleted


async def get_tasks_by_entity(entity_type: str, entity_id: int) -> List[Any]:
    """Get all tasks for a specific entity."""
    return await find_tasks_by_entity(entity_type, entity_id)


async def get_task_statuses() -> List[Any]:
    """Get all task statuses."""
    return await find_task_statuses()


async def get_task_priorities() -> List[Any]:
    """Get all task priorities."""
    return await find_task_priorities()


async def _build_entity_url(taskable_type: str, taskable_id: int) -> Tuple[Optional[str], Optional[str]]:
    """Build URL and lead_type for an entity. Returns (url, lead_type)."""
    url_map = {
        "Project": f"/projects/{taskable_id}",
        "Deal": f"/leads/deals/{taskable_id}",
        "Individual": f"/accounts/individuals/{taskable_id}",
        "Organization": f"/accounts/organizations/{taskable_id}",
    }

    if taskable_type in url_map:
        return url_map[taskable_type], None

    if taskable_type == "Lead":
        lead_type = await get_lead_type(taskable_id)
        if not lead_type:
            return None, None
        lead_type_plural = "opportunities" if lead_type == "Opportunity" else lead_type.lower() + "s"
        return f"/leads/{lead_type_plural}/{taskable_id}", lead_type

    if taskable_type == "Account":
        account_entity = await get_account_entity(taskable_id)
        if not account_entity:
            return None, None
        entity_type, entity_id = account_entity
        entity_path_map = {"Individual": "individuals", "Organization": "organizations"}
        path = entity_path_map.get(entity_type)
        if not path:
            return None, None
        return f"/accounts/{path}/{entity_id}", None

    return None, None


async def resolve_entity_display(taskable_type: Optional[str], taskable_id: Optional[int]) -> Dict[str, Any]:
    """Resolve entity display info for a task."""
    empty_result = {
        "type": None,
        "id": None,
        "display_name": None,
        "url": None,
        "lead_type": None,
    }

    if not taskable_type or not taskable_id:
        return empty_result

    display_name = await get_entity_display_name(taskable_type, taskable_id)
    url, lead_type = await _build_entity_url(taskable_type, taskable_id)

    return {
        "type": taskable_type,
        "id": taskable_id,
        "display_name": display_name,
        "url": url,
        "lead_type": lead_type,
    }


async def batch_resolve_entity_display(tasks: List[Any]) -> Dict[Tuple[str, int], Dict[str, Any]]:
    """Batch resolve entity display info for multiple tasks. Returns dict mapping (type, id) -> entity_info."""
    if not tasks:
        return {}
    
    # Collect unique entity pairs
    entity_pairs: List[Tuple[str, int]] = []
    for task in tasks:
        if task.taskable_type and task.taskable_id:
            pair = (task.taskable_type, task.taskable_id)
            if pair not in entity_pairs:
                entity_pairs.append(pair)
    
    if not entity_pairs:
        return {}
    
    # Batch fetch display names
    display_names = await batch_get_entity_display_names(entity_pairs)
    
    # Collect lead IDs and account IDs for URL building
    lead_ids: List[int] = []
    account_ids: List[int] = []
    for entity_type, entity_id in entity_pairs:
        if entity_type == "Lead":
            lead_ids.append(entity_id)
        elif entity_type == "Account":
            account_ids.append(entity_id)
    
    # Batch fetch lead types and account entities
    lead_types = await batch_get_lead_types(lead_ids) if lead_ids else {}
    account_entities = await batch_get_account_entities(account_ids) if account_ids else {}
    
    # Build result map
    result_map: Dict[Tuple[str, int], Dict[str, Any]] = {}
    empty_result = {
        "type": None,
        "id": None,
        "display_name": None,
        "url": None,
        "lead_type": None,
    }
    
    url_map = {
        "Project": lambda tid: f"/projects/{tid}",
        "Deal": lambda tid: f"/leads/deals/{tid}",
        "Individual": lambda tid: f"/accounts/individuals/{tid}",
        "Organization": lambda tid: f"/accounts/organizations/{tid}",
    }
    
    for entity_type, entity_id in entity_pairs:
        display_name = display_names.get((entity_type, entity_id))
        
        # Build URL
        url = None
        lead_type = None
        
        if entity_type in url_map:
            url = url_map[entity_type](entity_id)
        elif entity_type == "Lead":
            lead_type = lead_types.get(entity_id)
            if lead_type:
                lead_type_plural = "opportunities" if lead_type == "Opportunity" else lead_type.lower() + "s"
                url = f"/leads/{lead_type_plural}/{entity_id}"
        elif entity_type == "Account":
            account_entity = account_entities.get(entity_id)
            if account_entity:
                acc_entity_type, acc_entity_id = account_entity
                entity_path_map = {"Individual": "individuals", "Organization": "organizations"}
                path = entity_path_map.get(acc_entity_type)
                if path:
                    url = f"/accounts/{path}/{acc_entity_id}"
        
        result_map[(entity_type, entity_id)] = {
            "type": entity_type,
            "id": entity_id,
            "display_name": display_name,
            "url": url,
            "lead_type": lead_type,
        }
    
    return result_map

File: modules/agent-go/internal/services/task_service.go
package services

import (
	"errors"
	"fmt"
	"strings"

	"github.com/shopspring/decimal"

	"github.com/pina-colada-co/agent-go/internal/repositories"
	"github.com/pina-colada-co/agent-go/internal/schemas"
	"github.com/pina-colada-co/agent-go/internal/serializers"
)

var ErrTaskNotFound = errors.New("task not found")
var ErrTaskTitleRequired = errors.New("title is required")

// TaskService handles task business logic
type TaskService struct {
	taskRepo *repositories.TaskRepository
}

// NewTaskService creates a new task service
func NewTaskService(taskRepo *repositories.TaskRepository) *TaskService {
	return &TaskService{taskRepo: taskRepo}
}

// GetTasks returns paginated tasks
func (s *TaskService) GetTasks(page, pageSize int, orderBy, order, search, scope string, tenantID *int64, taskableType *string, taskableID *int64, projectID *int64) (*serializers.PagedResponse, error) {
	params := repositories.NewPaginationParams(page, pageSize, orderBy, order)
	params.Search = search

	// Only apply project filter when scope is "project"
	var effectiveProjectID *int64
	if scope == "project" && projectID != nil {
		effectiveProjectID = projectID
	}

	result, err := s.taskRepo.FindAll(params, tenantID, taskableType, taskableID, effectiveProjectID)
	if err != nil {
		return nil, err
	}

	items := make([]serializers.TaskResponse, len(result.Items))
	for i := range result.Items {
		entityInfo := s.resolveEntityInfo(result.Items[i].TaskableType, result.Items[i].TaskableID)
		items[i] = serializers.TaskToResponseWithEntity(&result.Items[i], entityInfo)
	}

	resp := serializers.NewPagedResponse(items, result.TotalCount, result.Page, result.PageSize, result.TotalPages)
	return &resp, nil
}

// GetTask returns a task by ID
func (s *TaskService) GetTask(id int64) (*serializers.TaskResponse, error) {
	task, err := s.taskRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if task == nil {
		return nil, errors.New("task not found")
	}

	entityInfo := s.resolveEntityInfo(task.TaskableType, task.TaskableID)
	resp := serializers.TaskToResponseWithEntity(task, entityInfo)
	return &resp, nil
}

// DeleteTask deletes a task
func (s *TaskService) DeleteTask(id int64) error {
	task, err := s.taskRepo.FindByID(id)
	if err != nil {
		return err
	}
	if task == nil {
		return ErrTaskNotFound
	}

	return s.taskRepo.Delete(id)
}

// CreateTask creates a new task
func (s *TaskService) CreateTask(input schemas.TaskCreate, tenantID *int64, userID int64) (*serializers.TaskResponse, error) {
	if input.Title == "" {
		return nil, ErrTaskTitleRequired
	}

	repoInput := repositories.TaskCreateInput{
		TenantID:               tenantID,
		UserID:                 userID,
		Title:                  input.Title,
		Description:            input.Description,
		TaskableType:           input.TaskableType,
		TaskableID:             input.TaskableID,
		CurrentStatusID:        input.CurrentStatusID,
		PriorityID:             input.PriorityID,
		Complexity:             input.Complexity,
		AssignedToIndividualID: input.AssignedToIndividualID,
	}

	if input.SortOrder != nil {
		repoInput.SortOrder = *input.SortOrder
	}
	if input.StartDate != nil {
		repoInput.StartDate = parseDate(input.StartDate)
	}
	if input.DueDate != nil {
		repoInput.DueDate = parseDate(input.DueDate)
	}
	if input.EstimatedHours != nil {
		d := decimal.NewFromFloat(*input.EstimatedHours)
		repoInput.EstimatedHours = &d
	}
	if input.ActualHours != nil {
		d := decimal.NewFromFloat(*input.ActualHours)
		repoInput.ActualHours = &d
	}

	taskID, err := s.taskRepo.Create(repoInput)
	if err != nil {
		return nil, err
	}

	return s.GetTask(taskID)
}

// UpdateTask updates an existing task
func (s *TaskService) UpdateTask(id int64, input schemas.TaskUpdate, userID int64) (*serializers.TaskResponse, error) {
	task, err := s.taskRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if task == nil {
		return nil, ErrTaskNotFound
	}

	updates := buildTaskUpdates(input, userID)
	if len(updates) == 0 {
		entityInfo := s.resolveEntityInfo(task.TaskableType, task.TaskableID)
		resp := serializers.TaskToResponseWithEntity(task, entityInfo)
		return &resp, nil
	}

	if err := s.taskRepo.Update(id, updates); err != nil {
		return nil, err
	}

	return s.GetTask(id)
}

// GetTasksByEntity returns tasks for a specific entity
func (s *TaskService) GetTasksByEntity(entityType string, entityID int64, tenantID *int64) ([]serializers.TaskResponse, error) {
	tasks, err := s.taskRepo.FindByEntity(entityType, entityID, tenantID)
	if err != nil {
		return nil, err
	}
	result := make([]serializers.TaskResponse, len(tasks))
	for i := range tasks {
		entityInfo := s.resolveEntityInfo(tasks[i].TaskableType, tasks[i].TaskableID)
		result[i] = serializers.TaskToResponseWithEntity(&tasks[i], entityInfo)
	}
	return result, nil
}

func buildTaskUpdates(input schemas.TaskUpdate, userID int64) map[string]interface{} {
	updates := map[string]interface{}{
		"updated_by": userID,
	}

	if input.Title != nil {
		updates["title"] = *input.Title
	}
	if input.Description != nil {
		updates["description"] = *input.Description
	}
	if input.TaskableType != nil {
		updates["taskable_type"] = *input.TaskableType
	}
	if input.TaskableID != nil {
		updates["taskable_id"] = *input.TaskableID
	}
	if input.CurrentStatusID != nil {
		updates["current_status_id"] = *input.CurrentStatusID
	}
	if input.PriorityID != nil {
		updates["priority_id"] = *input.PriorityID
	}
	if input.StartDate != nil {
		updates["start_date"] = parseDate(input.StartDate)
	}
	if input.DueDate != nil {
		updates["due_date"] = parseDate(input.DueDate)
	}
	if input.EstimatedHours != nil {
		updates["estimated_hours"] = decimal.NewFromFloat(*input.EstimatedHours)
	}
	if input.ActualHours != nil {
		updates["actual_hours"] = decimal.NewFromFloat(*input.ActualHours)
	}
	if input.Complexity != nil {
		updates["complexity"] = *input.Complexity
	}
	if input.SortOrder != nil {
		updates["sort_order"] = *input.SortOrder
	}
	if input.CompletedAt != nil {
		updates["completed_at"] = parseDate(input.CompletedAt)
	}
	if input.AssignedToIndividualID != nil {
		updates["assigned_to_individual_id"] = *input.AssignedToIndividualID
	}

	return updates
}

// resolveEntityInfo resolves entity display info for a task
func (s *TaskService) resolveEntityInfo(taskableType *string, taskableID *int64) *serializers.TaskEntityInfo {
	if taskableType == nil || taskableID == nil {
		return nil
	}

	tType := *taskableType
	tID := *taskableID

	displayName := s.taskRepo.GetEntityDisplayName(tType, tID)
	url, leadType := s.buildEntityURL(tType, tID)

	return &serializers.TaskEntityInfo{
		Type:        &tType,
		ID:          &tID,
		DisplayName: displayName,
		URL:         url,
		LeadType:    leadType,
	}
}

// buildEntityURL builds URL and lead_type for an entity
func (s *TaskService) buildEntityURL(taskableType string, taskableID int64) (*string, *string) {
	urlMap := map[string]string{
		"Project":      fmt.Sprintf("/projects/%d", taskableID),
		"Deal":         fmt.Sprintf("/leads/deals/%d", taskableID),
		"Individual":   fmt.Sprintf("/accounts/individuals/%d", taskableID),
		"Organization": fmt.Sprintf("/accounts/organizations/%d", taskableID),
	}

	if url, ok := urlMap[taskableType]; ok {
		return &url, nil
	}

	if taskableType == "Lead" {
		return s.buildLeadURL(taskableID)
	}

	if taskableType == "Account" {
		return s.buildAccountURL(taskableID)
	}

	return nil, nil
}

func (s *TaskService) buildLeadURL(taskableID int64) (*string, *string) {
	leadType := s.taskRepo.GetLeadType(taskableID)
	if leadType == nil {
		return nil, nil
	}

	leadTypePlural := pluralizeLeadType(*leadType)
	url := fmt.Sprintf("/leads/%s/%d", leadTypePlural, taskableID)
	return &url, leadType
}

func pluralizeLeadType(leadType string) string {
	if leadType == "Opportunity" {
		return "opportunities"
	}
	return strings.ToLower(leadType) + "s"
}

func (s *TaskService) buildAccountURL(taskableID int64) (*string, *string) {
	entityType, entityID := s.taskRepo.GetAccountEntity(taskableID)
	if entityType == nil || entityID == nil {
		return nil, nil
	}

	pathMap := map[string]string{
		"Individual":   "individuals",
		"Organization": "organizations",
	}

	path, ok := pathMap[*entityType]
	if !ok {
		return nil, nil
	}

	url := fmt.Sprintf("/accounts/%s/%d", path, *entityID)
	return &url, nil
}
